src/optimistic-oracle/optimistic-oracle.js

```diff
@@ -0,0 +1,232 @@
+
+// load agent configuration
+const config = require('../../agent-config.json');
+
+// provide ABI for ERC-20 decimals() function
+const erc20Abi = [
+  {
+    constant: true,
+    inputs: [],
+    name: 'decimals',
+    outputs: [
+      {
+        name: '',
+        type: 'uint8',
+      },
+    ],
+    payable: false,
+    stateMutability: 'view',
+    type: 'function',
+  },
+];
+
+const CHAIN_ID = 1; // mainnet
+
+// stores the optimistic oracle contract address
+let optimisticOracleAddress;
+
+// create ethers interface object
+const optimisticOracleAbi = getAbi('OptimisticOracle');
+const iface = new ethers.utils.Interface(optimisticOracleAbi);
+
+// set up provider for read-only contract queries
+const provider = new ethers.providers.JsonRpcProvider(getJsonRpcUrl());
+
+/*
+Events we are interested in:
+
+    event RequestPrice(
+        address indexed requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes ancillaryData,
+        address currency,
+        uint256 reward,
+        uint256 finalFee
+    );
+
+    event ProposePrice(
+        address indexed requester,
+        address indexed proposer,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes ancillaryData,
+        int256 proposedPrice,
+        uint256 expirationTimestamp,
+        address currency
+    );
+*/
+const eventNames = ['RequestPrice', 'ProposePrice'];
+
+// calculate the % difference between 2 values
+function calculatePercentError(first, second) {
+  const delta = first.minus(second).absoluteValue();
+  return delta.div(first).multipliedBy(100);
+}
+
+// get an ERC-20 token price using the CoinGecko API
+// free tier is limited to 50 calls/minute
+// see https://coingecko.com/en/api/documentation for more information
+async function getPrice(contractAddress) {
+  const baseUrl = 'https://api.coingecko.com/api/v3/simple/token_price/';
+  const id = 'ethereum';
+  const vsCurrency = 'usd';
+  // contractAddress must be checksum encoded
+  const requestUrl = `${baseUrl + id}?contract_addresses=${contractAddress}&vs_currencies=${vsCurrency}`;
+
+  const response = await axios.get(requestUrl);
+
+  // check that we got a valid response
+  if (response.status !== 200) {
+    throw new Error(`Error getting response from CoinGecko (status=${response.status})`);
+  }
+
+  // check that we got data for this token address
+  if (Object.keys(response.data).length === 0) {
+    throw new Error(`CoinGecko returned no data for token '${contractAddress}'`);
+  }
+
+  // in the response data, the token address key is NOT checksum encoded
+  const tokenAddress = contractAddress.toLowerCase();
+
+  // check that the currency denomination we requested is supported
+  if (!(Object.prototype.hasOwnProperty.call(response.data[tokenAddress], vsCurrency))) {
+    throw new Error(`CoinGecko has no data for token '${tokenAddress}' vs currency '${vsCurrency}'`);
+  }
+
+  const price = response.data[tokenAddress][vsCurrency];
```

Comment:
This may be a bit of a nit:
It was a bit surprising to me that the price is passed back as a string rather than a BigNumber.  I can see how preserving it as a string is useful for the `Request Price` event, because no calculations or comparisons are made, the string is just included in the alert.  However, my expectation based on the function name (`getPrice`) and variable name (`price`) was that a number would be returned, not a string representation of a number.

In the case of the `Propose Price` event, we call `getPrice` and then immediately cast the value to a `BigNumber` but use the same variable name, `price` as a string and then `price` as a BigNumber.

My suggestion would be to consider using `priceString` when it is a string and `price` (or `priceBigNumber`) when it is a number.

Reply:
I've changed this so that everything is a number.  I still do the conversion to BN to use the calculatePercentError() method that was reused from another agent, but I use different variable names to make it clear that the type is different.